### PR TITLE
Colt Parser

### PR DESCRIPTION
**Pull request by pke11y**
Initial commit for the Colt parser. Adds new CSV type base parser.

WIP: further tests to be added once #56 is merged. Currently this is required to ensure e2e email test is successful.

Includes individual tests for each attachment - iCal and CSV.

Adds `email` type to the e2e tests using `.eml` file with `.ics` and `.csv` attachments.

### CODE REVIEW

**Comment by chadell** [comment]
@pke11y good work! going on the right direction!

File: circuit_maintenance_parser/parser.py [hunk 1]
```diff
@@ -161,3 +161,30 @@ def clean_line(line):
...
+    _data_types = ["application/csv", "text/csv", "application/octet-stream"]
+
+    def parse(self, raw: bytes) -> List[Dict]:
+        """Execute parsing."""
+        result = []
+
+        data_base: Dict[str, Union[int, str, Iterable]] = {}
+        try:
+            for data in self.parse_csv(raw, data_base):
+                result.append(data)
```
Comment (chadell) [inline circuit_maintenance_parser/parser.py:178]: even it's not necessary for this parser... I can foresee that maybe some provider could send mutliple maintenances on a CSV notification, so it's ok to leave as it is

File: tests/unit/test_providers.py [hunk 2]
```diff
@@ -10,6 +10,7 @@
 from circuit_maintenance_parser.provider import (
     GenericProvider,
     Cogent,
+    Colt,
```
Comment (chadell) [inline tests/unit/test_providers.py:13]: I know this was the pattern before, but this is gonna change, so better if you do not update this file, less conflicts to fix :)

File: circuit_maintenance_parser/parser.py (hunk 1 above)
Comment (pke11y) [reply circuit_maintenance_parser/parser.py:178]: yes, I thought better to be safe for now.

File: tests/unit/test_providers.py (hunk 2 above)
Comment (pke11y) [reply tests/unit/test_providers.py:13]: Ah ok, I see the other PR now. Will remove.

**Comment by chadell** [comment]
#56 PR is merged ;)

File: circuit_maintenance_parser/parsers/colt.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    account = ""
+
+                    # organizer = (
+                    #     str(component.get("ORGANIZER")) if component.get("ORGANIZER") else self.default_organizer
+                    # )
+                    # provider = (
+                    #     str(component.get("X-MAINTNOTE-PROVIDER"))
+                    #     if component.get("X-MAINTNOTE-PROVIDER")
+                    #     else self.default_provider
+                    # )
```
Comment (chadell) [inline circuit_maintenance_parser/parsers/colt.py:55]:
```suggestion
```

File: circuit_maintenance_parser/parsers/colt.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                    summary_match = re.search(
+                        r"^.*?[-]\s(?P<maintenance_id>CRQ[\S]+).*?,\s*(?P<account>\d+)$", str(component.get("SUMMARY"))
+                    )
+                    if summary_match:
+                        maintenance_id = summary_match.group("maintenance_id")
+                        account = summary_match.group("account")
+
+                    data = {
+                        "circuits": [],
+                        # "provider": provider,
```
Comment (chadell) [inline circuit_maintenance_parser/parsers/colt.py:65]:
```suggestion
```

File: circuit_maintenance_parser/parsers/colt.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                        "account": account,
+                        "maintenance_id": maintenance_id,
+                        "status": Status("CONFIRMED"),
+                        "start": round(component.get("DTSTART").dt.timestamp()),
+                        "end": round(component.get("DTEND").dt.timestamp()),
+                        "stamp": round(component.get("DTSTAMP").dt.timestamp()),
+                        "summary": str(component.get("SUMMARY")),
+                        # "organizer": organizer,
+                        # "uid": str(component.get("UID")),
```
Comment (chadell) [inline circuit_maintenance_parser/parsers/colt.py:74]:
```suggestion
```

File: circuit_maintenance_parser/parsers/colt.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                        "start": round(component.get("DTSTART").dt.timestamp()),
+                        "end": round(component.get("DTEND").dt.timestamp()),
+                        "stamp": round(component.get("DTSTAMP").dt.timestamp()),
+                        "summary": str(component.get("SUMMARY")),
+                        "sequence": int(component.get("SEQUENCE")),
+                    }
+                    result.append(data)
+
+        except ValidationError as exc:
+            raise MissingMandatoryFields from exc
```
Comment (chadell) [inline circuit_maintenance_parser/parsers/colt.py:69]:
```suggestion
```

**Review by chadell — DISMISSED** [review]
LGTM

File: circuit_maintenance_parser/parsers/colt.py (hunk 6 above)
Comment (chadell) [reply circuit_maintenance_parser/parsers/colt.py:69]: this is useless now, as the `Maintenance` are not instantiated here

File: circuit_maintenance_parser/parsers/colt.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                        "maintenance_id": maintenance_id,
+                        "status": Status("CONFIRMED"),
+                        "start": round(component.get("DTSTART").dt.timestamp()),
+                        "end": round(component.get("DTEND").dt.timestamp()),
+                        "stamp": round(component.get("DTSTAMP").dt.timestamp()),
+                        "summary": str(component.get("SUMMARY")),
+                        "sequence": int(component.get("SEQUENCE")),
+                    }
+                    result.append(data)
+
```
Comment (chadell) [inline circuit_maintenance_parser/parsers/colt.py:67]:
```suggestion
```

File: circuit_maintenance_parser/parsers/colt.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
+"""Colt parser."""
+
+import logging
+import re
+import csv
+import io
+import base64
+
+from pydantic import ValidationError
```
Comment (chadell) [inline circuit_maintenance_parser/parsers/colt.py:9]:
```suggestion
```

File: circuit_maintenance_parser/parsers/colt.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+import logging
+import re
+import csv
+import io
+import base64
+
+from icalendar import Calendar  # type: ignore
+
+from circuit_maintenance_parser.parser import Csv
+from circuit_maintenance_parser.errors import ParserError, MissingMandatoryFields
```
Comment (chadell) [inline circuit_maintenance_parser/parsers/colt.py:12]:
```suggestion
from circuit_maintenance_parser.errors import ParserError
```

File: circuit_maintenance_parser/parsers/colt.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+                    summary_match = re.search(
+                        r"^.*?[-]\s(?P<maintenance_id>CRQ[\S]+).*?,\s*(?P<account>\d+)$", str(component.get("SUMMARY"))
+                    )
+                    if summary_match:
+                        maintenance_id = summary_match.group("maintenance_id")
+                        account = summary_match.group("account")
+
+                    data = {
+                        "circuits": [],
```
Comment (chadell) [inline circuit_maintenance_parser/parsers/colt.py:55]:
```suggestion
```

File: tests/unit/data/colt/colt1_result.json [hunk 11]
```diff
@@ -0,0 +1,13 @@
+[
+  {
+    "account": "12345000",
+    "circuits": [],
```
Comment (chadell) [inline tests/unit/data/colt/colt1_result.json:4]:
```suggestion
```

**Pull request merged**